### DISCUSSION TITLE
esp32s3 + st7789 no image

### DISCUSSION THREAD

**Discussion by cheops** [lvgl-micropython/lvgl_micropython#17] [Q&A] [answered]
Hello,

I have a board with esp32s3 + st7789  
The board and display works with latest micropython and [st7789_mpy](https://github.com/russhughes/st7789_mpy)
But I fail to get any image with lvgl_micropython.
Any help or ideas are appreciated.
Thanks

lvgl_micropython compilation:
```
python3 make.py esp32 mpy_cross submodules clean LV_CFLAGS="-DLV_COLOR_DEPTH=16" BOARD=ESP32_GENERIC_S3 BOARD_VARIANT=SPIRAM_OCT DISPLAY=st7789
```
lvgl_micropython test script
```
import lcd_bus
import st7789
from micropython import const
import lvgl as lv
import task_handler
import gc

_MOSI = const(6)
_MISO = const(8)
_SCLK = const(7)
_CS = const(5)
_DC = const(4)
_RESET = const(48)
_FREQ = const(10_000_000)
_WIDTH = const(296)
_HEIGHT = const(240)

bus = lcd_bus.SPIBus(
    dc=_DC,
    host=1,
    mosi=_MOSI,
    miso=_MISO,
    sclk=_SCLK,
    cs=_CS,
    freq=_FREQ
)

gc.collect()

display=st7789.ST7789(
    data_bus=bus,
    display_width=_WIDTH,
    display_height=_HEIGHT,
    reset_pin=_RESET,
    reset_state=st7789.STATE_LOW,
    color_byte_order=st7789.BYTE_ORDER_BGR,
    color_space=lv.COLOR_FORMAT.RGB565
)

# display.set_rotation(lv.DISPLAY_ROTATION._90)
display.init()
lv.init()
task_handler.TaskHandler()

btn=lv.button(lv.screen_active())
btn.set_size(100,100)
lbl=lv.label(btn)
lbl.set_text("Press me, I'm a button")
btn.center()

label2 = lv.label(lv.screen_active())
label2.set_long_mode(lv.label.LONG.SCROLL_CIRCULAR) # Circular scroll
label2.set_width(150)
label2.set_text("It is a circularly scrolling text. ")
label2.align(lv.ALIGN.CENTER, 0, 40)
```

as reference, this is the script I use with bare micropython and st7789_mpy
```
# ESP32 Example
# To use baudrates above 26.6MHz you must use my firmware or modify the micropython
# source code to increase the SPI baudrate limit by adding SPI_DEVICE_NO_DUMMY to the
# .flag member of the spi_device_interface_config_t struct in the machine_hw_spi_init_internal.c
# file.  Not doing so will cause the ESP32 to crash if you use a baudrate that is too high.

# https://github.com/russhughes/st7789_mpy

import machine
import time
import st7789
import gc

mosi_pin=machine.Pin(6)
miso_pin=machine.Pin(8)
sck_pin=machine.Pin(7)

spi = machine.SPI(2, baudrate=80_000_000, sck=sck_pin, mosi=mosi_pin, miso=miso_pin)

reset_pin=machine.Pin(48, machine.Pin.OUT)
dc_pin=machine.Pin(4, machine.Pin.OUT)
cs_pin=machine.Pin(5, machine.Pin.OUT)

rotations = [(st7789.MADCTL_MV | st7789.BGR, 296, 240, 0, 0)]

gc.collect()

display = st7789.ST7789(spi=spi,
                        width=296,
                        height=240,
                        rotation=0,
                        rotations=rotations,
                        reset=reset_pin,
                        dc=dc_pin,
                        cs=cs_pin,
                        inversion=False,
                        buffer_size=296*240*2
                        )

display.init()
sleeptime=0.5

for _ in range(10):
    display.fill(st7789.RED)
    time.sleep(sleeptime)

    display.fill(st7789.GREEN)
    time.sleep(sleeptime)

    display.fill(st7789.BLUE)
    time.sleep(sleeptime)

    display.fill(st7789.YELLOW)
    time.sleep(sleeptime)

display.fill(st7789.BLACK)
```

**Comment by cheops** [comment]
How would I start debugging this?
Where do I add extra code?
Can I write to the framebuffers directly (a full red screen for example)?

**Comment by kdschlosser** [comment]
I have to update the example code because things have changed. I forgot to update the example and I am sorry for that. 

I just had back surgery this past Tuesday (2 days ago) and I have not spent much time sitting at the computer for obvious reasons. I have popped in for a little bit here and there but that's about it. haven't wanted to sit for any kind of a long amount of time due to the quite large amount of pain i am in. 

I had a hole drilled through one of my vertebrae to have some material removed from around my spinal cord and around one of the branches for the sciatic nerve that runs down my right leg.  Typical healing time is 12 to 16 weeks. 

I know you really want things to move forward with this but unfortunately it is going to have to wait a bit longer until I am able to really dive back in. I was out of commission for almost 6 weeks before the surgery and that has made the progress really slow with this project. I am really sorry but to be sure this is not something I planned to have happen nor did I want to have happen.

**Reply by cheops** [reply]
I followed your story already a bit. I am really sorry to hear that.  
The operation sounds quite terrible and I wish you a swift recovery.  
I definitely don't want to put any pressure on you, and want to give you all the time you need.  

I will wade a bit more through the code, any working example that somebody has is welcome.

Wishing you all the best!

**Comment by kdschlosser** [comment]
```python
from micropython import const


_MOSI = const(6)
_MISO = const(8)
_SCLK = const(7)
_CS = const(5)
_DC = const(4)
_RESET = const(48)
_FREQ = const(80_000_000)
_WIDTH = const(296)
_HEIGHT = const(240)


import lcd_bus

bus = lcd_bus.SPIBus(
    dc=_DC,
    host=2,
    mosi=_MOSI,
    miso=_MISO,
    sclk=_SCLK,
    cs=_CS,
    freq=_FREQ
)


import st7789
import lvgl as lv


display = st7789.ST7789(
    data_bus=bus,
    display_width=_WIDTH,
    display_height=_HEIGHT,
    reset_pin=_RESET,
    reset_state=st7789.STATE_LOW,
    color_space=lv.COLOR_FORMAT.RGB565,
    rgb565_byte_swap=True
)


# display.set_rotation(lv.DISPLAY_ROTATION._90)
display.init()

btn = lv.button(lv.screen_active())
btn.set_size(100, 100)
lbl = lv.label(btn)
lbl.set_text("Press me, I'm a button")
btn.center()

label2 = lv.label(lv.screen_active())
label2.set_long_mode(lv.label.LONG.SCROLL_CIRCULAR) # Circular scroll
label2.set_width(150)
label2.set_text("It is a circularly scrolling text. ")
label2.align(lv.ALIGN.CENTER, 0, 40)


import time


while True:
    time.sleep_ms(1)
    lv.tick_inc(1)
    lv.task_handler()

```

**Comment by cheops** [comment]
I updated to the latest version (25/03) and I tried this, but still the same result.
my display stays black, nothing is drawn.

**Comment by kdschlosser** [comment]
what display and MCU are you using?

**Reply by cheops** [reply]
ESP32-S3 16MB Flash 8MB PSRAM
ST7789 display 296x240 pixels with rounded corners, a bit like this [link](https://www.taobao.com/list/item/678831948508.htm)

**Comment by kdschlosser** [comment]
Need to know who makes the display. I say this because you specifically said "my display stays black" which makes me wonder if there is a backlight pin that you are not using and maybe there is a defaulted on in that other library that is being used or perhaps you passed a backlight pin to that other library and just forgot about it.

I am going to add to the ESP32 build using custom partition sizes because MicroPython doesn't have proper support for the different flash sizes. It's OK with the memory just not with the partitions. Right now you are only able to use 8mb of that 16mb of storage.

**Reply by cheops** [reply]
Hello Kevin,  
I like your approach to start with the basics.  
The display is connected with a flat cable to a custom pcb.  
The backlight is connected to a hardware switch.   
I can see the backlight is on at the side of the screen (and can turn it off with the hardware switch).

The display is in reset, when the reset pin is pulled low.
the reset function of the st7789mpy library: [link](https://github.com/russhughes/st7789_mpy/blob/master/st7789/st7789.c#L335)

datasheet of the [display ](https://github.com/Fri3dCamp/badge_2024_hw/blob/main/Datasheets/QT020JA001-A0_SPEC.pdf) and [displaycontroller](https://github.com/Fri3dCamp/badge_2024_hw/blob/main/Datasheets/ST7789V.pdf)

**Reply by cheops** [reply]
difference is in the rotations settings, I will recheck with `st7789.MADCTL_MV | st7789.BGR` set in the `MADCTL` register

**Reply by cheops** [reply]
I rechecked with writing `MADCTL` = `0x28` , which corresponds to `st7789.MADCTL_MV | st7789.BGR`

still nothing on the display

**Comment by kdschlosser** [comment]
> The display is in reset, when the reset pin is pulled low.

I have not seen a display yet that needed to have the reset pin pulled low to reset it. Try commenting out the line for the reset pin state just to see what happens..

**Reply by cheops** [reply]
This is the hardware reset pin, software reset is also possible.
I verified the hardware reset (with the other firmware build), setting `.value(0)` blanks the display if there is anything on it, so my configuration seems correct  
leaving it out is also a viable option, then software reset is used.
```
    reset_pin=_RESET,
    reset_state=st7789.STATE_LOW,
```

I tried 3 options:
- no definition `reset_pin=None`
- `reset_state=st7789.STATE_LOW`
- `reset_state=st7789.STATE_HIGH`

none of the above 3 resulted in any visible image on the screen

So what is the next step to investigate?

**Reply by kdschlosser** [reply]
next step is me locating my SPI display that has the same IC on on it and seeing if I can get it running.

**Reply by cheops** [reply]
ow sorry, if I can add local debugging to check things, let me know

**Reply by cheops** [reply]
I found this:
https://github.com/lvgl/lv_binding_micropython/blob/master/driver/generic/st77xx.py#L424
```
            ## FIXME: needs out of sleep mode AGAIN, otherwise will stay bleck the first time on?
            (ST77XX_SLPOUT, None, 100),
```
added it at the end of `st7789.py` and hey, I see random bits on the display (no more BLACK screen)

but calling something like this does not update the display
```
arr = bytearray(b'\x77' * 2 * 296 * 1)
display._set_memory_location(0, 0, 295, 1)
bus.tx_color(0x2c, arr, 0, 0, 295, 1)
```

calling `display.invert_colors()` does change the random pattern of noise on the screen

if I add
```
import task_handler  # NOQA
th = task_handler.TaskHandler()
```
I can see from the debug output in `lcd_types.c:161` that esp_lcd_panel_io_tx_color gets called, but the display image does not update

**Comment by cheops** [comment, marked as answer]
Hello Kevin, we get an image!
![image](https://github.com/kdschlosser/lvgl_micropython/assets/4328487/9c5348ba-db77-4924-b162-94daa3c0baa3)

git commit version cabd909
```
from micropython import const

MOSI = const(6)
MISO = const(8)
SCLK = const(7)
CS = const(5)
DC = const(4)
RESET = const(48)
FREQ = const(80_000_000)
WIDTH = const(240)
HEIGHT = const(296)

import lcd_bus

bus = lcd_bus.SPIBus(
    dc=DC,
    host=1,
    mosi=MOSI,
    miso=MISO,
    sclk=SCLK,
    cs=CS,
    freq=FREQ
)

import st7789
import lvgl as lv
lv.init()

buf1 = bus.allocate_framebuffer(14_208, lcd_bus.MEMORY_DMA)
buf2 = bus.allocate_framebuffer(14_208, lcd_bus.MEMORY_DMA)

display = st7789.ST7789(
    data_bus=bus,
    display_width=WIDTH,
    display_height=HEIGHT,
    frame_buffer1=buf1,
    frame_buffer2=buf2,
    reset_pin=RESET,
    reset_state=st7789.STATE_LOW,
    color_space=lv.COLOR_FORMAT.RGB565,
    color_byte_order=st7789.BYTE_ORDER_RGB,
    rgb565_byte_swap=True
)
display.set_rotation(lv.DISPLAY_ROTATION._90)
display.init()


import task_handler  # NOQA
th = task_handler.TaskHandler()

screen = lv.screen_active()
screen.set_style_bg_color(lv.color_hex(0x777777), 0)

slider = lv.slider(screen)
slider.set_size(200, 50)
slider.center()
slider.set_value(50, 0)

label = lv.label(screen)
label.set_text('Hello Kevin, we get an image!')
label.align_to(slider, lv.ALIGN.OUT_TOP_MID, 0, -10)
```

**Reply by kdschlosser** [reply]
Yeah I figured out what the issue was. It was exceedingly stupid problem too. I had marked the dc pin object as an object instead of an integer. simple thing that I kept on overlooking. It wasn't until I added debugging output was the problem seen....

**Comment by marilantagliari** [comment]
Is it possible to use with ST7701 ??

**Reply by kdschlosser** [reply]
I don't have that driver added.. I will add the driver right now.

**Reply by kdschlosser** [reply]
This is going to be complicated to add. This is because of the display supporting RGB  and DSI interfaces. The setup of the display is done using a 3 wire SPI connection. The reason why it is tricky is because the SPI needs to be a bit-bang implementation because of the crazy low speeds the SPI runs at. The hardware SPI is not able to communicate at the low speed needed.

In order for me to add this I need to add a new module written in C code that will add the SPI functionality. I then have to extend the current display driver framework to accept this second connection type to the display so commands can be sent using it. I should be able to share 3 of the data wires for the RGB display bus so additional pins on the ESP32 don't need to be used. RGB displays don't typically need to have command sent to them while they are in operation and that is what will allow me to do this. However... If you want to have hardware screen rotation working all the time then you will have to use a separate set of pins to allow this. IO Expanders are also sometimes used to provide the additional pins needed for the SPI. This would not be supported right away as it would take me a bit to think of how to make that all work properly.

I can add the display driver and the additional driver needed for the SPI no issues there. It is going to take me a bit to get it added as it is not as simple as writing a python driver for a display using the display framework I have already written. Expect it to take me a bit to get running as it is going to be something that gets put to the end of the list of things I already have to do. I have a bunch of bugs I need to sort out beforehand,

**Reply by marilantagliari** [reply]
thanks a lot for the answer kdchlosser.. excellent your code... I'm searching same solution to use this display but not success until now. I don't need rotation... or may I need to buy another display board.. this display is a part of my solution to monitor boats, so I'm thinking to do in Flutter with Rasp3, but I don't like Rasp for embedded application.

But now I have an issue executing LVGL_MICROPYTHON in ESP32S3... 

`MicroPython a61c446-dirty on 2024-07-09; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3
Type "help()" for more information.
>>> import lvgl as lv
>>> lv.init()
>>> s = lv.obj()

A fatal error occurred. The crash dump printed below may be used to help
determine what caused it. If you are not already running the most recent
version of MicroPython, consider upgrading. New versions often fix bugs.

To learn more about how to debug and/or report this crash visit the wiki
page at: https://github.com/micropython/micropython/wiki/ESP32-debugging

MPY version : a61c446-dirty on 2024-07-09
IDF version : 11eaf41b
Machine     : Generic ESP32S3 module with Octal-SPIRAM with ESP32S3

Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x42030659  PS      : 0x00060830  A0      : 0x8202dcdb  A1      : 0x3fcbd550  
A2      : 0x00000000  A3      : 0x00000000  A4      : 0x00000000  A5      : 0x3fcbd84c  
A6      : 0x00000001  A7      : 0x3fcbd550  A8      : 0x00000000  A9      : 0x3fcbd530  
A10     : 0x3c254550  A11     : 0x00000038  A12     : 0x00000000  A13     : 0x3fcbd84c  
A14     : 0x00000001  A15     : 0x3fcbd530  SAR     : 0x00000002  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000036  LBEG    : 0x400570e8  LEND    : 0x400570f3  LCOUNT  : 0x00000000  


Backtrace: 0x42030656:0x3fcbd550 0x4202dcd8:0x3fcbd590 0x4202c8bf:0x3fcbd5d0 0x4201482e:0x3fcbd610 0x42006105:0x3fcbd650 0x420be92d:0x3fcbd690 0x420065a1:0x3fcbd6b0 0x42014876:0x3fcbd700 0x420bd471:0x3fcbd730 0x420be92d:0x3fcbd750 0x420be9f5:0x3fcbd770 0x40379aa2:0x3fcbd790 0x420b61bb:0x3fcbd830 0x420be92d:0x3fcbd880 0x420be942:0x3fcbd8a0 0x420cad93:0x3fcbd8c0 0x420caed6:0x3fcbd950 0x420d256e:0x3fcbd9a0




ELF file SHA256: 117aa061b

Rebooting...
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x18 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40382ce6
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3820,len:0x105c
load:0x403c9700,len:0x4
load:0x403c9704,len:0xbd8
load:0x403cc700,len:0x2e30
entry 0x403c989c
MicroPython a61c446-dirty on 2024-07-09; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3
Type "help()" for more information.
>>> `

This is the backtrace:

`MacBook-Air-de-Marilan:lvgl_micropython marilantagliari$ xtensa-esp32s3-elf-addr2line -pfiaC -e lib/micropython/ports/esp32/build-ESP32_GENERIC_S3-SPIRAM_OCT/micropython.elf 0x42030656:0x3fcbd550 0x4202dcd8:0x3fcbd590 0x4202c8bf:0x3fcbd5d0 0x4201482e:0x3fcbd610 0x42006105:0x3fcbd650 0x420be92d:0x3fcbd690 0x420065a1:0x3fcbd6b0 0x42014876:0x3fcbd700 0x420bd471:0x3fcbd730 0x420be92d:0x3fcbd750 0x420be9f5:0x3fcbd770 0x40379aa2:0x3fcbd790 0x420b61bb:0x3fcbd830 0x420be92d:0x3fcbd880 0x420be942:0x3fcbd8a0 0x420cad93:0x3fcbd8c0 0x420caed6:0x3fcbd950 0x420d256e:0x3fcbd9a0
0x42030656: lv_obj_mark_layout_as_dirty at ??:?
0x4202dcd8: lv_obj_class_init_obj at ??:?
0x4202c8bf: lv_obj_create at ??:?
0x4201482e: mp_lv_obj_create at lv_mp.c:?
0x42006105: lv_fun_builtin_var_call at lv_mp.c:?
0x420be92d: mp_call_function_n_kw at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/runtime.c:717
0x420065a1: make_new at lv_mp.c:?
0x42014876: obj_make_new at lv_mp.c:?
0x420bd471: type_call at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/objtype.c:1003
0x420be92d: mp_call_function_n_kw at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/runtime.c:717
0x420be9f5: mp_call_method_n_kw at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/runtime.c:733
0x40379aa2: mp_execute_bytecode at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/vm.c:1042
0x420b61bb: fun_bc_call at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/objfun.c:267
0x420be92d: mp_call_function_n_kw at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/runtime.c:717
0x420be942: mp_call_function_0 at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/py/runtime.c:691
0x420cad93: parse_compile_execute at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/shared/runtime/pyexec.c:123
0x420caed6: do_reader_stdin at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/shared/runtime/pyexec.c:280
 (inlined by) pyexec_raw_repl at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/shared/runtime/pyexec.c:513
0x420d256e: mp_task at /Users/marilantagliari/dev_app/LCD_Tracker/lvgl_micropython/lib/micropython/ports/esp32/main.c:157`

any suggestion?
Did you thing this is an error related with display driver or an issue with compile environment ?

**Reply by kdschlosser** [reply]
open up a new issue for this instead of hijacking this conversation please
